### PR TITLE
Merge hapi, a high-level API with traning and inference into Paddle.

### PR DESCRIPTION
**Pull request by qingqing01**
Hapi is a high level API for training and inference.
The main modules include Model, Loss, Metrics, Dataset.
Also includes some common modules and models in NLP and computer vision, such as BERT, ResNet.

These modules are developed by:
0YuanZhang0, guoshengCS heavengate, LielinJiang, qingqing01, xyzhou-puck huangjun12, wangxiao1021 zhangyang

There are 3 PRs to merge:  first is core framework,  other two are vision and text module.  This is the first PR.

tree dir is : 
```
`-- hapi
    |-- __init__.py
    |-- callbacks.py
    |-- datasets
    |   |-- __init__.py
    |   |-- flowers.py
    |   |-- folder.py
    |   |-- mnist.py
    |   `-- utils.py
    |-- distributed.py
    |-- download.py
    |-- logger.py
    |-- loss.py
    |-- metrics.py
    |-- model.py
    |-- progressbar.py
    |-- tests
    |-- tests
    |   |-- CMakeLists.txt
    |   |-- test_model.py
    |   |-- ...
    |-- utils.py
    `-- vision
        |-- __init__.py
        `-- models
            |-- __init__.py
            `-- lenet.py
```

### CODE REVIEW

File: python/paddle/incubate/hapi/progressbar.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import sys
+import time
+import numpy as np
+from collections import namedtuple
```
Comment (luotao1) [inline python/paddle/incubate/hapi/progressbar.py:23]:
```
[10:54:33][Step 1/1] ----Exception in get api filename----
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] paddle.incubate.hapi.progressbar.namedtuple module is collections
```
可以加一个`__all__ = ['xxx']`试试

Comment (luotao1) [reply python/paddle/incubate/hapi/progressbar.py:23]:
```
[10:54:33][Step 1/1] ----Exception in get api filename----
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] paddle.distributed.fs_wrapper.PurePosixPath.with_suffix module is pathlib
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] ----Exception in get api filename----
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] paddle.distributed.fs_wrapper.PurePosixPath module is pathlib
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] ----Exception in get api filename----
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] paddle.distributed.fs_wrapper.PurePosixPath.is_reserved module is pathlib
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] ----Exception in get api filename----
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] paddle.distributed.fs_wrapper.PurePosixPath.joinpath module is pathlib
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] ----Exception in get api filename----
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] 
[10:54:33][Step 1/1] paddle.distributed.fs_wrapper.PurePosixPath.is_absolute module is pathlib
```
可以在`fs_wrapper.py`里加`__all__ = ['xxx']`试试

Comment (lelelelelez) [reply python/paddle/incubate/hapi/progressbar.py:23]: 1. 此错误目前只是报错，不会影响ci合入。
2. 已经在[#24325 ](https://github.com/PaddlePaddle/Paddle/pull/24325)修复了文件名为空的bug。
3. 已经rerun了，报错内容可见CI的log。

**Comment by qingqing01** [comment]
```
[16:04:49]
[Step 1/1] ****************
[16:04:49]
[Step 1/1] 0. You must have Dianhai approval for change 20+ files or add than 1000+ lines of content.
[16:04:49]
[Step 1/1] 1. You must have one RD (XiaoguangHu01 or lanxianghit) and one TPM (saxon-zh or Boyan-Liu or swtkiwi) approval for the api change for the management reason of API interface.
[16:04:49]
[Step 1/1] 2. You must have one TPM (saxon-zh or Boyan-Liu or swtkiwi) approval for the api change for the management reason of API document.
[16:04:49]
[Step 1/1] 3. Unittest is not allowed to be disabled.
[16:04:49]
[Step 1/1] You must have one RD (kolinwei(Recommend), liuwei1031, or luotao1) approval for the usage of @unittest.skip or @unittest.skipIf.
[16:04:49]
[Step 1/1] +@unittest.skipIf(not fluid.is_compiled_with_cuda(),
[16:04:49]
[Step 1/1] +@unittest.skipIf(not fluid.is_compiled_with_cuda(),
[16:04:49]
[Step 1/1] +@unittest.skipIf(not fluid.is_compiled_with_cuda(),
```

@XiaoguangHu01  @lanxianghit @saxon-zh @Boyan-Liu  @swtkiwi  @kolinwei @luotao1 @liuwei1031  @raindrops2sea  Please help to review. Thanks!

**Review by kolinwei — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/incubate/hapi/metrics.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    Base class for metric, encapsulates metric logic and APIs
+    Usage:
+        
+        m = SomeMetric()
+        for prediction, label in ...:
+            m.update(prediction, label)
+        m.accumulate()
+        
+    Advanced usage for :code:`add_metric_op`
+    Metric calculating con be accelerate by calucateing metric states
```
Comment (Boyan-Liu) [inline python/paddle/incubate/hapi/metrics.py:45]: "Metric calculation can be accelerated by caculating metric"

Comment (Boyan-Liu) [inline python/paddle/incubate/hapi/metrics.py:45]: "Metric calculation can be accelerated by calculating metric"

**Review by qingqing01 — COMMENTED** [review]
@Boyan-Liu Thanks.   There is still other PR to submit,  we can fix it in next PR.

**Review by Boyan-Liu — APPROVED** [review]

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**